larevt/CalibrationDBI/Providers/DBDataset.h

```diff
@@ -0,0 +1,163 @@
+#ifndef DBDATASET_H
+#define DBDATASET_H
+//=================================================================================
+//
+// Name: DBDataset.h
+//
+// Purpose: Header for class DBDataset.
+//          This class represents data extracted from the opaque wda struct Dataset,
+//          which struct represents the result of a calibration database query on a 
+//          particular database table in an IOV database.  Data is extracted and 
+//          copied from the Dataset struct using the wda api.
+//
+//          Database data are essentially a rectangular array of values, indexed by
+//          (row, column).  Accessors are provided to access data as string, long, 
+//          or double.
+//
+//          Rows are labeled by channel number.  Columns are labeled by name and type.
+//
+//          Binary values, which are stored as strings by wda, are stored as
+//          std::variant<long, double, unique_ptr<std::string> > > in this class.
+//
+//          Columns are labeled by column name and type.
+//
+// Data members:
+//
+// fBeginTime - IOV begin validity time.
+// fEndTime   - IOV end validity time.
+// fNRows     - Number of rows.
+// fNCols     - Number of columns.
+// fColNames  - Names of columns.
+// fColTypes  - Data types of columns.
+// fChannels  - Channel numbers (indexed by row number).
+// fData      - Calibration data.
+//
+// Normally, the first element of each row is an integer channel number.
+// Furthermore, it can be assumed that rows are ordered by increasing channel number.
+// 
+// Calibration data is contained in data member fData, which is a rectangular array 
+// of strings of dimension fNRows x fNCols.  For efficiency, calibration data are
+// stored using a single std::vector<std::string>, which array is allocated at
+// construction or update time.  Elements are accessed columnwise as follows.
+//
+// value = fData[fNCols*row + column]
+//
+// Or use the provided accessors.
+//
+// Nested class DBRow provides access to data from a single database row.
+//
+// Created: 26-Oct-2020 - H. Greenlee
+//
+//=================================================================================
+
+#include <string>
+#include <vector>
+#include <variant>
+#include <memory>
+#include "larevt/CalibrationDBI/Interface/CalibrationDBIFwd.h"
+#include "larevt/CalibrationDBI/IOVData/IOVTimeStamp.h"
+
+namespace lariov
+{
+  class DBDataset
+  {
+
+  public:
+
+    // Typedef
+
+    typedef std::variant<long, double, std::unique_ptr<std::string> > value_type;
```

Comment:
The `std::unique_ptr<std::string>` type will always take more memory than simply `std::string` as it adds another layer of indirection through a pointer--it's an additional 8 bytes.  Switching to `std::variant<long, double, std::string>` has the effect of increasing the stack size of `std::variant<...>`, but in so doing it shifts what was heap memory (via the unique_ptr indirection) to stack memory.